### PR TITLE
[JIT SSA] Added testing for the Cat Op in LazyTensor

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76694
* __->__ #76552
* #76703

### CODE REVIEW

File: test/cpp/lazy/test_symbolic_shape.cpp [hunk 1]
```diff
@@ -127,7 +127,32 @@ TEST_F(LazyShapeTest, TestMulBasic) {
...
+  EXPECT_EQ(getIsSymbolic(res), expected);
+
+  // Test handling of symbolic dims of inequal sizes, Currently crashes
+  // As we can't handle cases where upper bound dims are not equal
+  /*
+  torch::Tensor e = tensorWithSymbolicShape({2, 2}, {true, false});
+  torch::Tensor f = tensorWithSymbolicShape({2, 3}, {false, true});
+  res = torch::cat({e, f}, 0);
+  expected = {true, false};
+  EXPECT_EQ(getIsSymbolic(res), expected);
```
Comment (Gamrix) [inline test/cpp/lazy/test_symbolic_shape.cpp:153]: @Krovatkin  This part of the test is currently commented out, as the LazyTensor shape function for `torch.cat` errors out with an error saying that the dim=1 does not match up. 
In the ideal implementation of upper bound shapes should run successfully in our case, as the upper bound should be assumed not to be exact. However, one can make the argument that this is a case that doesn't need to be handled. I am wondering: Should we actually fix this case, and if so, how should we do so.

Comment (Krovatkin) [reply test/cpp/lazy/test_symbolic_shape.cpp:153]: okay this is a bit alarming since this may potentially mean we need the whole universe of parallel shape functions for every op. In theory, we could easily run into the same issue with any binary op where two dimensions (neither is 1-dim) won't line up since one of them is a conservative estimate.

In practice, however, it seems to be less likely since such case would probably break in eager as well as  unless a user has some super natural ability to predict that data dependent ops would produce the exact shape that `cat` or some other op would expect.

wdy guys t @wconstab @miladm @JackCaoG  ?

Comment (wconstab) [reply test/cpp/lazy/test_symbolic_shape.cpp:153]: > LazyTensor shape function for torch.cat errors out with an error saying that the dim=1 does not match up.
what's the precise scenario/call-flow?
({2,2} and {2,3})  are these upper bounds or are these real shapes?  In either case, they don't match, so I don't see how it's valid to feed them to cat.  Please elaborate on the logic for why this is valid in some sense.

Comment (Gamrix) [reply test/cpp/lazy/test_symbolic_shape.cpp:153]: [I brought this up in the LazyTensor meeting today, our conclusion is that for now, we should ignore this case]

Comment (Gamrix) [reply test/cpp/lazy/test_symbolic_shape.cpp:153]: To answer your question though, the values are upper bounds.

**Review by Krovatkin — APPROVED** [review]
:shipit:
